cpp/src/arrow/engine/query_plan.h

```diff
@@ -0,0 +1,99 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
```

Comment:
I don't think that work on query planning is anticipated soon, only work on execution, so this could be put off later.

As a nit, I would also suggest calling this "logical_plan.h" to distinguish logical from physical execution.

Reply:
Ok, perhaps the naming is bad, but this is not a logical plan. This is the physical plan. "Exec plan" is a particular instantiation of executing this physical plan.

Does that make sense?

Reply:
I see. How about:

* query_plan.h -> physical_plan.h
* exec_plan.h -> exec_node.h